### ISSUE TITLE
Error when assigning `Set[str]` to a variable of type `Set[Union[str, int]]`

### ISSUE THREAD

**Issue by mauvilsa** [python/mypy#14068]
**Bug Report**

For a variable with type `Set[str]` naturally has a type which is a subset of a more general type such as `Set[Union[str, int]]`. However, mypy shows an error when assigning.

**To Reproduce**

```python
from typing import Set, Union

var1: Set[Union[str, int]] = {'a', 'b'}
var2: Set[str] = {'a', 'b'}
var1 = var2
```

https://mypy-play.net/?mypy=latest&python=3.10&gist=22bafdb063089ecdcacbd01eb32dc76d

**Expected Behavior**

Don't show any error.

**Actual Behavior**

```
main.py:5: error: Incompatible types in assignment (expression has type "Set[str]", variable has type "Set[Union[str, int]]")
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 0.990
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.8, 3.10 in mypy-play.net

**Comment by AlexWaygood** [comment]
This is expected behavior due to the fact that `set` is an [invariant collection](https://mypy.readthedocs.io/en/stable/generics.html#variance-of-generics).

A workaround is to use `typing.AbstractSet` instead — due to the fact that `AbstractSet` is immutable, it is a covariant class, unlike `set`:

```python
from typing import AbstractSet, Union

var1: AbstractSet[Union[str, int]] = {'a', 'b'}
var2: AbstractSet[str] = {'a', 'b'}
var1 = var2
```

**Comment by mauvilsa** [comment]
@AlexWaygood thank you for the response. Though this leads to more questions. What is the need for `typing.Set` being invariant? I see that `typing.AbstractSet` is [deprecated](https://docs.python.org/3/library/typing.html#typing.AbstractSet) suggesting to use `collections.abc.Set`. Isn't this confusing and unintuitive?

**Comment by JelleZijlstra** [comment]
`Set` is invariant for the same reason `List` is, as explained at the link Alex posted.

Yes, `typing.AbstractSet` is soft deprecated in favor of `collections.abc.Set`. If you think the name is confusing, I don't disagree.

**Comment by AlexWaygood** [comment]
> What is the need for `typing.Set` being invariant?

For very similar reasons to the ones given in the docs I linked to above for why `List` is invariant.

> I see that `typing.AbstractSet` is [deprecated](https://docs.python.org/3/library/typing.html#typing.AbstractSet) suggesting to use `collections.abc.Set`. Isn't this confusing and unintuitive?

Yeah, `AbstractSet` was deprecated along with all the other aliases in the `typing` module as part of PEP 585. But the name of `collections.abc.Set` is _really_ unfortunate :(

It would be great if we could go back in time and call it `AbstractSet`. Unfortunately it's existed for much longer than `typing.AbstractSet` has, so there's not really much chance of the name being changed now. I usually do `from collections.abc import Set as AbstractSet` in my own code.
